### PR TITLE
Issue 3576: Improve logging to make it easier to debug security issues

### LINKED ISSUE #3576: Improve security-related logging
**Problem description**
Here are a few problems with security logging that needs to be addressed immediately. There might be more, but the focus in this issue is on the following: 

1. On the client side, if you enable logging, a message shows up at a `WARN` level: ` The credentials are not specified or could not be extracted`. This happens whether `Auth` is enabled or not. 

    Moreover, even when `Auth` is enabled and the credentials are actually specified, this message shows up upon invoking `StreamManager.create(clientConfigWithCredentialsSpecified)`. 

   It's rather confusing (and even annoying) to see this message. 

2. Currently, there is no logging in Controller and Segment store service that tells whether auth and TLS are enabled. This can be problematic while troubleshooting issues, such as when security is enabled in one service and not the other. 

**Problem location**
* For 1: ClientConfig.ClientConfigBuilder.build()
* For 2: Classes ControllerServiceStarter and ServiceStarter

**Suggestions for an improvement**
* Remove the unnecessary warning message from ClientConfig.ClientConfigBuilder.build()
* Add logging about whether auth and TLS are enabled in ControllerServiceStarter and ServiceStarter

### PR DESCRIPTION
**Pull request by ravisharda**
**Change log description**  
The PR addresses two things:
* Removes a log entry about missing credentials in ClientConfig which was erroneous and misleading (as explained in issue #3576)
* Logs controller and segment store configuration (including security configuration). For sensitive security configuration, it only tells whether the configuration is specified. 

**Purpose of the change**  
Resolves #3576 

**What the code does**  
* Removes a log entry in ClientConfig that was causing problem no. 1 mentioned in the issue. 
* Adds custom toString() implementations to Controller and Segment Store configuration classes where security information present; for rest of the config classes, adds Lombok toString annotation. 
 
   While the scope of the PR was logging for security debugging, it didn't make sense to log only security configuration; The code would be much less elegant and would add a bit of inessential complexity if we limit the scope to logging security config only. So, all configuration is logged, which will also be useful for debugging in general. 
* Uses the above config objects' `toString()` methods to log the config when the Controller and Segment Store Services are started. 

While the code logs the config, it avoids logging configuration information that is sensitive from a security perspective. 

Below are a few examples of what gets logged: 

*Controller Server Config:*
```
ControllerServiceConfigImpl(threadPoolSize=80, storeClientConfig=StoreClientConfigImpl(storeType=Zookeeper, 
zkClientConfig=Optional[ZKClientConfigImpl(connectionString: localhost:4000, namespace: 
pravega/singlenode-92105678-0ece-4afb-a44c-bc0d64d424eb, initialSleepInterval: 2000, maxRetries: 1,
 sessionTimeoutMs: 10000, secureConnectionToZooKeeper: true, trustStorePath is specified, 
trustStorePasswordPath is specified)]), 
hostMonitorConfig=io.pravega.controller.store.host.impl.HostMonitorConfigImpl@638465ac, 
controllerClusterListenerEnabled=false, 
timeoutServiceConfig=TimeoutServiceConfig(maxLeaseValue=120000), 
eventProcessorConfig=Optional[ControllerEventProcessorConfigImpl(scopeName=_system, 
commitStreamName=_commitStream, 
commitStreamScalingPolicy=ScalingPolicy(scaleType=FIXED_NUM_SEGMENTS, targetRate=0, 
scaleFactor=0, minNumSegments=2), abortStreamName=_abortStream, 
abortStreamScalingPolicy=ScalingPolicy(scaleType=FIXED_NUM_SEGMENTS, targetRate=0, 
scaleFactor=0, minNumSegments=2), scaleStreamName=_requeststream, 
scaleStreamScalingPolicy=ScalingPolicy(scaleType=FIXED_NUM_SEGMENTS, targetRate=0, 
scaleFactor=0, minNumSegments=2), commitReaderGroupName=commitStreamReaders, 
commitReaderGroupSize=1, abortReaderGroupName=abortStreamReaders, abortReaderGroupSize=1, 
scaleReaderGroupName=scaleGroup, scaleReaderGroupSize=1, 
commitCheckpointConfig=CheckpointConfig(type=Periodic, 
checkpointPeriod=CheckpointConfig.CheckpointPeriod(numEvents=10, numSeconds=10)), 
abortCheckpointConfig=CheckpointConfig(type=Periodic, 
checkpointPeriod=CheckpointConfig.CheckpointPeriod(numEvents=10, numSeconds=10)), 
scaleCheckpointConfig=CheckpointConfig(type=None, checkpointPeriod=null))], 
gRPCServerConfig=Optional[GRPCServerConfigImpl(port: 9090, publishedRPCHost: localhost, 
publishedRPCPort: 9090, authorizationEnabled: true, userPasswordFile is specified, tokenSigningKey is 
specified, tlsEnabled: true, tlsCertFile is specified, tlsKeyFile is specified, tlsTrustStore is specified, 
replyWithStackTraceOnError: false, requestTracingEnabled: true)], 
restServerConfig=Optional[RESTServerConfigImpl(host: 0.0.0.0, port: 9091, tlsEnabled: true, keyFilePath is 
specified, keyFilePasswordPath is specified)])

```

*SegmentStore ServiceConfig:*
```
ServiceConfig(containerCount: 4, coreThreadPoolSize: 20, storageThreadPoolSize: 200, 
listeningPort: 6000, listeningIPAddress: 127.0.1.1, publishedPort: 6000, publishedIPAddress: 127.0.1.1,
zkURL: localhost:4000, zkRetrySleepMs: 5000, zkSessionTimeoutMs: 10000, zkRetryCount: 5, 
secureZK: true, zkTrustStore is specified, zkTrustStorePasswordPath is specified, 
clusterName: singlenode-0ca69158-a58a-48ac-a6e2-b7169f23203b, dataLogTypeImplementation: 
INMEMORY, storageImplementation: INMEMORY, storageImplementation: INMEMORY, 
readOnlySegmentStore: false, enableTls: true, certFile is specified, keyFile is specified, cachePolicy is 
specified, replyWithStackTraceOnError: false, instanceId: )
```

*SegmentStore AutoScalerConfig:*
```
AutoScalerConfig(controllerUri: tls://localhost:9090, internalRequestStream: _requeststream, 
cooldownDuration: PT10M, muteDuration: PT10M, cacheExpiry: PT20M, cacheCleanup: PT5M, 
tlsEnabled: true, tlsCertFile is specified, authEnabled: true, tokenSigningKey is specified, 
validateHostName: false, threadPoolSize: 10)
```

**How to verify it**  
All unit and system tests should pass.

### CODE REVIEW

File: client/src/main/java/io/pravega/client/ClientConfig.java [hunk 1]
```diff
@@ -94,9 +94,6 @@ public ClientConfig build() {
                 controllerURI = URI.create("tcp://localhost:9090");
             }
             extractCredentials();
-            if (credentials == null) {
```
Comment (fpj) [inline client/src/main/java/io/pravega/client/ClientConfig.java]: It seems fine and even desirable to remove this message, but one question I have is whether we are warning the user when security is disabled. We should have a WARN message to make sure a user is not running with security disabled by accident.

Comment (ravisharda) [reply client/src/main/java/io/pravega/client/ClientConfig.java]: If the server is running with `auth` enabled, the client application sees an exception indicating authentication failure in both these cases: 
* Credentials are invalid 
* Credentials are missing

So, the client will get an exception if the credentials are missed by accident. (On a different note, the indication - i.e., the exception name and the message - has significant room for improvement, but I'll leave that for another PR.)

The problem that we are addressing here is that the logs on the client side indicate that credentials are missing even in these cases: 
* The client application has actually supplied them (due to some internal calls that happen under the hoods in the client library). 
* The client application is talking to a cluster where `auth` is disabled. In real applications, `auth` should be enabled, and therefore it shouldn't be a real issue in those cases. But, when using a cluster for development/testing or a standalone mode cluster, application developers can get confused by this message. 

I've faced these issues and have got confused on several occasions. I've also seen people pointing to this message when something else has gone wrong elsewhere. :)

File: segmentstore/server/host/src/main/java/io/pravega/segmentstore/server/host/stat/AutoScalerConfig.java [hunk 2]
```diff
@@ -131,4 +132,28 @@ private AutoScalerConfig(TypedProperties properties) throws ConfigurationExcepti
     public static ConfigBuilder<AutoScalerConfig> builder() {
         return new ConfigBuilder<>(COMPONENT_CODE, AutoScalerConfig::new);
     }
+
+    @Override
+    public String toString() {
+        // Note: We don't use Lombok @ToString to automatically generate an implementation of this method,
+        // in order to avoid returning a string containing sensitive security configuration.
+
+        return new StringBuilder("AutoScalerConfig(")
```
Comment (fpj) [inline segmentstore/server/host/src/main/java/io/pravega/segmentstore/server/host/stat/AutoScalerConfig.java]: According to codecov, this is the code that is causing a low code coverage mark.

Comment (ravisharda) [reply segmentstore/server/host/src/main/java/io/pravega/segmentstore/server/host/stat/AutoScalerConfig.java]: Sure, Flavio. I've addressed this by adding a few tests, in the latest commit that I just pushed.

**Comment by ravisharda** [comment]
Just realized that there are additional entry points and configuration objects where logging is required. Additional changes are on their way.

**Comment by ravisharda** [comment]
> Just realized that there are additional entry points and configuration objects where logging is required. Additional changes are needed.

I've added these changes in the latest commit. The PR is ready for further review.

**Pull request merged**